### ISSUE TITLE
Graceful shutdown

### ISSUE THREAD

**Issue by StefanNemeth** [prisma/prisma#2917]
## Problem
I'd like to shutdown my application gracefully and do some database work before I exit the process myself.

However, Prisma exits the process on a SIGINT event on its own. Even though I got a work-around by monkey-patching "process.exit", the Prisma query engine still exits with the SIGINT event and I'm unable to execute queries therefore.

`The Node.js process already received a SIGINT signal, therefore the Prisma query engine exited
and your request can't be processed.`

## Suggested solution
1. Make prisma stop listening to process events by setting an option
2. Stop the engines manually by executing a stop method (?) just after the application is ready to exit 

## Alternatives
here's the very hacky work-around that I use (though I'd appreciate something else)

```javascript
const realProcessExit = process.exit;
process.exit = () => {
  // Monkey-patch process.exit because Prisma wants to exit before
  // we can shut down gracefully
};

import { PrismaClient } from "@prisma/client";

["EXIT", "SIGINT", "SIGUSR1", "SIGUSR2", "uncaughtException", "SIGTERM"].forEach((eventType) => {
  const previousListeners = process.listeners(eventType);
  process.removeAllListeners(eventType);

  process.once(eventType, async (param) => {
    await shutdown();

    // Make prisma stop the engines after the application is ready to exit
    previousListeners.forEach((l) => l(param));

    if (eventType !== "EXIT") {
      realProcessExit();
    }
  });
});
```

**Comment by JonathanCa97** [comment]
I discovered the same problem today

**Comment by KevinDenys** [comment]
Any news about this issue, I have the same problem

**Comment by timsuchanek** [comment]
Thanks for reporting @StefanNemeth!
Your use-case makes a lot of sense and we should definitely support it.
Unfortunately there doesn't seem to be a mechanism for running the exit hooks in order.
If such a thing would exist, we could run the disconnect after your cleanup would have run.

I also don't see a way right now to remove our exit even listeners, as for "ordinary" users, this provides safety to make sure, that the query engine and database connection is stopped properly, because we can't rely on Node.js there.

Therefore, your suggestion about disabling the listeners makes sense to me.

Suggestion for configuration:

```ts
new PrismaClient({
  disconnectOnExit: false
})
```

Would this solve your issue @StefanNemeth ?

Does this make sense to add @matthewmueller @Jolg42 ?

**Comment by matthewmueller** [comment]
> Cleans up spawned processes when the parent process dies.

@timsuchanek it looks like [execa](https://github.com/sindresorhus/execa/) deals with this same problem.

I did a quick search through their codebase and found the [signal-exit](https://github.com/tapjs/signal-exit) module. I didn't spend enough time tracing the code, but I think this is what they use to cleanup a zombie child.

Using the reproduction you provided me.

**index.js**

```js
const { spawn } = require('child_process')
const path = require('path')
const child = spawn('node', [path.join(__dirname, 'run.js')], {
  stdio: 'inherit',
})
setTimeout(() => {
  process.exit()
}, 100)
```

**run.js**

```js
console.log('hao', process.pid)

setTimeout(() => {
  console.log('hi')
}, 100000000)
```

It appears that adding `signal-exit` to **index.js** will work:

```js
const { spawn } = require('child_process')
const onExit = require('signal-exit')
const path = require('path')
const child = spawn('node', [path.join(__dirname, 'run.js')], {
  stdio: 'inherit',
})

onExit(function (code, signal) {
  console.log('killing child!')
  child.kill()
})

onExit(function (code, signal) {
  console.log('killing child2!')
})

setTimeout(() => {
  console.log('done')
}, 5000)
```

Two neat things about `signal-exit` from my testing:

1. You don't need to call process.exit() yourself. This may fix the problem described in this issue already.
2. We can recommend using this module to hook into the exit process. It calls call `onExit` callbacks. 

One limitation is these callbacks must run within the same tick, so you couldn't have a setTimeout inside them. I think this should work for most cases though.

**Comment by timsuchanek** [comment]
The suggested solution with `disconnectOnExit: false` however has the disadvantage, that child processes might get leaked.

Would a "before exit" hook like this work for you?

```
const prisma = new PrismaClient()

prisma.on('beforeExit', async () => {
  await prisma.user.deleteMany()
})
```

**Comment by StefanNemeth** [comment]
> The suggested solution with `disconnectOnExit: false` however has the disadvantage, that child processes might get leaked.
> 
> Would a "before exit" hook like this work for you?
> 
> ```ts
> const prisma = new PrismaClient()
> 
> prisma.on('beforeExit', async () => {
>   await prisma.user.deleteMany()
> })
> ```

This would definitely work. Nonetheless, I'd personally prefer the disconnectOnExit approach because I don't think that it should be the responsibility of a dependency (Prisma) to execute the shutdown logic of the entire application.

**Comment by timsuchanek** [comment]
Thanks for the feedback @StefanNemeth!
The reason Prisma is cleaning up its child processes, is this example:


`index.js`
```ts
const { spawn } = require('child_process')
const path = require('path')
const child = spawn('node', [path.join(__dirname, 'run.js')], {
  stdio: 'inherit',
})
setTimeout(() => {
  process.exit()
}, 100)
```
`run.js`
```ts
console.log('hao', process.pid)
setTimeout(() => {
  console.log('hi')
}, 100000000)
```

In this case, we'll end up with a dangling child process.
So unfortunately Node.js does not make sure, that the child process is killed properly.
That's why Prisma Client is cleaning this up, for now, to not introduce a foot gun for Prisma users.

**Comment by dfee** [comment]
The opt-out and callback on exit seems to make sense though. I just discovered that the reason my NestJs app isn't performing it's `onApplicationShutdown` cleanups are that Prisma is killing the process before it gets there.

I know this is slatted for current development, so wanted to get my thoughts in. Thanks!

**Comment by timsuchanek** [comment]
Thanks for all the helpful comments!
We just fixed it and introduced the `prisma.$on('beforeExit', () => ...)` hook.
While we're still killing the child processes, after the `beforeExit` hook has been executed, we're not killing the current process anymore, if there are other listeners on the signals, like SIGINT.

The reason is, that if you provide a SIGINT listener yourself, you are responsible for exiting the process yourself anyway already.

This way you have more control. It should fix the issue you just described @dfee and also the one mentioned in https://github.com/prisma/prisma/issues/3773

**Comment by microwavekonijn** [comment]
I apologize for reacting to a closed issue, but I have to say I find the solution backward. It basically says we don't trust you for closing the client properly and you will end up with a leak. Let us disable the behavior and give us an example of how to do it properly.

The solution I have now is to stop Prisma from exiting by letting it wait on a promise in the `beforeExit` listener, and call the resolve of said promise once NestJS reaches the `onApplicationShutdown` hook of the Prisma client.

**Comment by matthewmueller** [comment]
Hey @microwavekonijn, could you clarify the problem you're having? 

To clarify the current behavior as I understand it, when the process receives a SIGINT:

1. `beforeExit` hooks are called _(if defined)_
2. Child processes (query engine) also receive the SIGINT and presumably exit
3. The node process will exit if there's no SIGINT handler, if there is a SIGINT handler, that handler will be called.

Does this behavior match your understanding? Do you have a problem with this behavior?

**Comment by microwavekonijn** [comment]
I use NextJS to manage the life cycle of my application. I want to start Prisma in the `onModuleInit` hook and close it in the `OnApplicationShutdown` hook. Before that shutdown hook is called other 'onApplicationShutdown` and `onModuleDestory` can be called, if I want to execute any queries there I can't as I have to add them to the `beforeExit` hook from Prisma. NestJS is the one that figures things out like what order to execute the hooks, but with the `beforeExit` hook this is not possible(or at least not easily done).

The simplest solution for me would be to disable the SIGINT listener from Prisma and when the application closes when it is Prisma's turn I tell it to `$disconnect` and wait for it to clean up. It is nice to have a helping hand to avoid issues like leaks, but it can also greatly complicate things.

**Comment by matthewmueller** [comment]
Would listening for SIGINT _only_ if there's a `beforeExit` handler work for you? Then if there's no `beforeExit` defined, it doesn't listen for SIGINT?

**Comment by RafaelKr** [comment]
> Would listening for SIGINT _only_ if there's a `beforeExit` handler work for you? Then if there's no `beforeExit` defined, it doesn't listen for SIGINT?

Hi Matthew! Would this even work? How does the query engine know when to exit then?

**Comment by microwavekonijn** [comment]
All I want is to call `$disconnect` myself even in the case of a SIGINT, that way I have completely predictable behavior for my app.

I looked at the code and found that I can just returned a never resolving promise in the callback for `beforeExit`, blocking the SIGINT behavior.
